src/session/session-utils.ts

```diff
@@ -105,3 +112,49 @@ export function sessionEqual(
   copyB.sort(([k1], [k2]) => (k1 < k2 ? -1 : 1));
   return JSON.stringify(copyA) === JSON.stringify(copyB);
 }
+
+export function createGetJwtSessionId(config: ConfigInterface) {
+  return (shop: string, userId: string): string => {
+    return `${createSanitizeShop(config)(shop, true)}_${userId}`;
+  };
+}
+
+export function createGetOfflineId(config: ConfigInterface) {
+  return (shop: string): string => {
+    return `offline_${createSanitizeShop(config)(shop, true)}`;
+  };
+}
+
+export function createGetCurrentSessionId(config: ConfigInterface) {
+  return async ({
+    request,
+    isOnline,
+  }: GetCurrentSessionIdParams): Promise<string | undefined> => {
+    if (config.isEmbeddedApp) {
+      const authHeader = request.headers.authorization;
+      if (authHeader) {
+        const matches = (
+          typeof authHeader === 'string' ? authHeader : authHeader[0]
+        ).match(/^Bearer (.+)$/);
+        if (!matches) {
+          throw new ShopifyErrors.MissingJwtTokenError(
+            'Missing Bearer token in authorization header',
+          );
+        }
+
+        const jwtPayload = await createDecodeSessionToken(config)(matches[1]);
+        const shop = jwtPayload.dest.replace(/^https:\/\//, '');
+        if (isOnline) {
+          return createGetJwtSessionId(config)(shop, jwtPayload.sub);
+        } else {
+          return createGetOfflineId(config)(shop);
```

Comment:
Yeah, I was thinking about this too. The silver lining is that usually this will only apply to the "building block" functions like utils, which tend to be small and so a bit less bloated.

Because there's no global scope, it's not very easy to have one function reference another internal one without a reference to the overarching object, but I'm not sure we want to pass that external reference around internally, it feels a bit error prone to me.